### ISSUE TITLE
digest in digest_set and ResourceDescriptor is not valid JSON

### ISSUE THREAD

**Issue by gerardborst** [in-toto/attestation#269]
We from [Argos Notary](https://github.com/argosnotary) would very much like to migrate from an in-toto like structure to using attestations.

The problem is, I think, that the [digest definition](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md) isn't valid JSON as defined in [json.org](https://www.json.org/json-en.html).

```
{
  "<ALGORITHM_1>": "<HEX/BASE64 VALUE>",
  "<ALGORITHM_2>": "<HEX/BASE64 VALUE>",
  ... 
}
```
A JSON object should consist of static keys with values.

This is also the case in `ResourceDescriptor`.

In our project we are using openapi, Jackson, Spring Data, MongoDB, Javascript, React, Axis. All these expect valid JSON and you get in a lot of trouble if it isn't.

As an example we are searching on hash in our database but in mongodb it isn't possible to index on this `digest` because the name key isn't static.

**Comment by gerardborst** [comment]
As explained in rfc8259:

> An object structure is represented as a pair of curly brackets
   surrounding zero or more name/value pairs (or members).  A name is a
   string.  A single colon comes after each name, separating the name
   from the value.

I interpret this as a name/key is static with the variable value.

**Comment by TomHennen** [comment]
Hey Gerard,

We talked about this in the maintainers meeting today.  We're not sure we agree this is a violation of the JSON spec, but we do get the impression you're having some trouble using the spec with some of these other systems.  Can you tell us more about the specific problems you're having?

**Comment by welcor** [comment]
I agree with @gerardborst, but I also see why you don't experience the problem - this is mostly an issue when you translate json to static types. The stack Gerard is using translates the JSON into objects as part of the workflow. 

When you receive an object with variable keys, you generally only have the option of parsing it as a generic map internally (either `Map<String, String>` or, in some cases, `Map<String, Object>`). This, in turn, makes it very hard to propagate in a smart manner to other systems.

A switch to a format like the one below makes it easy to build a type-safe object for.
```
"digests" : [
  { 
    "alg": "<alg1>",
    "digest": "<digest1>"
  }, {
    "alg": "<alg2>",
    "digest": "<digest2>"
  }
]
```
Such a type-safe object can be used for indices, filtering, etc. in a much more convenient manner.

**Comment by TomHennen** [comment]
Thanks for that feedback.

What's interesting is that the structure you propose may make indexing in some stacks easier, it makes usage in others harder.  E.g. in code users now have to iterate through a list of digests to find the one they want instead of just looking up the one they need in a map.

**Comment by adityasaky** [comment]
Possibly a slight hack and may require some infrequent changes to the static schema but can the enumerated algorithms listed in https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md be used as possible keys in the hash object?

**Comment by gerardborst** [comment]
For my recap. 

I really think that not all Python dictionaries can be translated to Javascript Objects as these are intended. As it is now you expose Python internals in a generic interface which should be language independent. The same for the `_type` property which is by Python convention a property only for internal use. It feels the most natural to use Python objects/classes instead of dictionaries.

> A collection of name/value pairs. In various languages, this is realized as an object, record, struct, dictionary, hash table, keyed list, or associative array.

I think there is a reason they start with object because that would be the least problematic implementation.

> E.g. in code users now have to iterate through a list of digests to find the one they want instead of just looking up the one they need in a map.

This depends completely on the implementation. In our case the objects are stored in a mongo database and with an index created on `alg` it will be O(log n). 

I also don't understand the use case of more than 1 hash for a digest...

> Note that there are now many existing implementations.

Is this really the case?

Soon I have a lot of time on my hand so I really wanted to help. I had prepared some examples to demonstrate the huge difference but I think there is no point. 

Argos Notary is in principal a closed system and I think we will make our own implementation just as we did for the original in-toto design.

**Comment by gerardborst** [comment]
I was afraid the [Envelop](https://github.com/secure-systems-lab/dsse/blob/v1.0.0/envelope.md) had the same issue but that's not the case.

This issue can be closed in my opinion.

**Comment by adityasaky** [comment]
@gerardborst since you're the producer and consumer of argos notary metadata, you control the selection of the hash algorithms too, yes? I urge you to consider indexing by the algorithms themselves in that case so that the metadata remains compatible with other in-toto implementations _and_ you get the indexing benefits.

>I also don't understand the use case of more than 1 hash for a digest...

The idea is to enable agility for when we need to transition away from an algorithm with known weaknesses to a stronger algorithm. It can also allow for interoperability between different systems that use different hashes.

**Comment by gerardborst** [comment]
@adityasaky thank you for thinking along. I think you proposal is workable. 

> The idea is to enable agility for when we need to transition away from an algorithm with known weaknesses to a stronger algorithm. It can also allow for interoperability between different systems that use different hashes.

For our implementation a transition to another algorithm would be a major undertaking because of the use case we are trying to support. In our use case the complete verification is initially only done during the release. After the release the documents are stored in a dedicated collection. This collection is queried by the real consumers. 

The consumers can query with a random set of file hashes, even from different packages, to check if these are delivered by a trusted supply chain and trusted organizations. So all digests in the database should have all possible hashes otherwise this will not work.

So if you want to transition from algorithm A to B at first all digests in the database should have hashes A and B. If this is the case the client implementations can change to algorithm B. I think this would take years but maybe still something to take into account.

Conclusion is, and you made me more aware, that we have to choose the initial algorithm more carefully.
And I will investigate your idea.

**Comment by marcelamelara** [comment]
@gerardborst Friendly ping. Are there any outstanding issues we may help you resolve on this?

**Comment by gerardborst** [comment]
Hi @marcelamelara ,

I first tried to use only sha256 but discovered that this was to inflexible. I decided after this to use [extension-fields](https://slsa.dev/spec/v1.0/provenance#extension-fields)
```
ArgosDigest:
      type: object
      required:
        - hash
        - hashAlgorithm
      properties:
        hash:
          $ref: '#/components/schemas/HashString'
        algorithm:
          $ref: '#/components/schemas/HashAlgorithm'
```

But I'm still not sure. 

I'm leaning to using in-toto and also SLSA just as inspiration and not to be compatible with them, at least for now.

I think this issue can be closed.
